src/udev/udev-builtin-path_id.c

```diff
@@ -671,6 +682,21 @@ static int builtin_path_id(sd_device *dev, int argc, char *argv[], bool test) {
                 udev_builtin_add_property(dev, test, "ID_PATH_TAG", tag);
         }
 
+        /*
+         * Compatible link generation for ATA devices
+         * we need to strip the characters after "ata-<port>"
+         * in the path and assign the value to env variable
+         * ID_PATH_ATA_COMPAT
+         */
+        p = strstr(path, "ata-");
+        if (p) {
```

Comment:
Can't say I like this part. In particular, we concatenate various substrings here to get this path, and if any of them contain "ata-", things are going to go very wrong.

I'd like to see the following approach instead: let `handle_scsi_iscsi` return two strings (`char **path, char **compat_path`, the second one may be NULL if not needed), and then let `handle_scsi` return those two strings too, and finally have `builtin_path_id` assign `ID_PATH_ATA_COMPAT` if that second string is non-null. This might be even less code than the current patch, and it'll be much less fragile.

Reply:
Thanks for the great suggestion. I suppose you mean handle_scsi_iscsi as 
handle_scsi_ata here. Since the whole path name is composed from subsys and sys name, I implemeneted it in two ways. 
1\ Only adding the optional parameter compat_path in handle_scsi_ata. I still have to get
compat_path from path as the last version. compat_path is only used to identify if it's really
from ATA devices.
2\ After adding the compat_path, I compose the compat_path from all possible sys names. 
Then in builtin_path_id, I only assign compat_path to ID_PATH_ATA_COMPAT if it's not 
NULL.
What I sent for review is the second one. Please share your opinion.  Thanks.